plutus-core-interpreter/test/DynamicBuiltins/Definition.hs

```diff
@@ -0,0 +1,137 @@
+-- | A dynamic built-in name test.
+
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module DynamicBuiltins.Definition
+    ( test_definition
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Constant.Dynamic
+import           Language.PlutusCore.Generators.Interesting
+import           Language.PlutusCore.MkPlc
+
+import           Language.PlutusCore.StdLib.Data.Bool
+import qualified Language.PlutusCore.StdLib.Data.Function   as Plc
+import qualified Language.PlutusCore.StdLib.Data.List       as Plc
+
+import           DynamicBuiltins.Common
+
+import           Data.Either                                (isRight)
+import           Data.Maybe
+import           Data.Proxy
+import           Hedgehog                                   hiding (Size, Var)
+import qualified Hedgehog.Gen                               as Gen
+import qualified Hedgehog.Range                             as Range
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+import           Test.Tasty.HUnit
+
+dynamicFactorialName :: DynamicBuiltinName
+dynamicFactorialName = DynamicBuiltinName "factorial"
+
+dynamicFactorialMeaning :: DynamicBuiltinNameMeaning
+dynamicFactorialMeaning = DynamicBuiltinNameMeaning sch fac where
+    sch =
+        TypeSchemeAllSize $ \s ->
+            TypeSchemeBuiltin (TypedBuiltinSized (SizeBound s) TypedBuiltinSizedInt) `TypeSchemeArrow`
+            TypeSchemeBuiltin (TypedBuiltinSized (SizeBound s) TypedBuiltinSizedInt)
+    fac n = product [1..n]
+
+dynamicFactorialDefinition :: DynamicBuiltinNameDefinition
+dynamicFactorialDefinition =
+    DynamicBuiltinNameDefinition dynamicFactorialName dynamicFactorialMeaning
+
+dynamicFactorial :: Term tyname name ()
+dynamicFactorial = dynamicBuiltinNameAsTerm dynamicFactorialName
+
+-- | Check that the dynamic factorial defined above computes to the same thing as
+-- a factorial defined in PLC itself.
+test_dynamicFactorial :: TestTree
+test_dynamicFactorial =
+    testCase "dynamicFactorial" $ do
+        let env = insertDynamicBuiltinNameDefinition dynamicFactorialDefinition mempty
+            lhs = typecheckEvaluateCek env $ applyFactorial dynamicFactorial 3 10
+            rhs = typecheckEvaluateCek mempty $ applyFactorial factorial 3 10
+        assertBool "type checks" $ isRight lhs
+        lhs @?= rhs
+
+dynamicConstName :: DynamicBuiltinName
+dynamicConstName = DynamicBuiltinName "const"
+
+dynamicConstMeaning :: DynamicBuiltinNameMeaning
+dynamicConstMeaning = DynamicBuiltinNameMeaning sch Prelude.const where
+    sch =
+        TypeSchemeAllType @"a" @0 Proxy $ \a ->
```

Comment:
I was horrified and then I remembered we mostly only need well-scopedness, so this is okay and we'll rename it later? Is that true? Do we rename the types of dynamic builtins when we introduce them during typechecking?

Reply:
> Do we rename the types of dynamic builtins when we introduce them during typechecking?

Yes. When turning `DynamicBuiltinNameMeanings` into `DynamicBuiltinNameTypes`, we normalize the types along the way using the safe interface of the type normalizer which renames its argument first.

We also always explicitly apply functions to those bound type variables and never inline them. E.g.

```haskell
instance KnownDynamicBuiltinType a => KnownDynamicBuiltinType (PlcList a) where
    toTypeEncoding _ = TyApp () (_recursiveType listData) $ toTypeEncoding (Proxy @a)
```

This ensures that a free variable won't become bound by an unexpected binder. And the renamer handles name shadowing issues.

The current assumption we have is that only `QuotedTerm` is allowed to have a non-closed type and all other `KnownDynamicBuiltinType` must be closed and uniqueness of type variables in `QuotedTerm` is managed by the user. This looks like a pretty safe design to me (modulo explicit uniques managing, but I'm going to fix that at some point).